src/Form/ActionForm.php

```diff
@@ -0,0 +1,188 @@
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides the form functions to call actions on referenced entities.
+ */
+class ActionForm implements ContainerInjectionInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountProxyInterface
+   */
+  protected $currentUser;
+
+  /**
+   * The messenger service.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * ActionForm constructor.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager service.
+   * @param \Drupal\Core\Session\AccountProxyInterface $currentUser
+   *   The current user.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, AccountProxyInterface $currentUser, MessengerInterface $messenger) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->currentUser = $currentUser;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static($container->get('entity_type.manager'), $container->get('current_user'), $container->get('messenger'));
+  }
+
+  /**
+   * Build the form elements.
+   *
+   * @param array $element
+   *   The element with the attached action form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   * @param array $context
+   *   The context of this form.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function buildForm(array &$element, FormStateInterface $form_state, array $context) {
+
+    /** @var \Drupal\Core\Field\FieldItemListInterface $items */
+    $items = $context['items'];
+    if ($items->isEmpty()) {
+      return;
+    }
+
+    $field_definition = $items->getFieldDefinition();
+
+    $entity_type = $field_definition->getSettings()['target_type'];
+
+    $actionStorage = $this->entityTypeManager->getStorage('action');
+    $actions = array_filter($actionStorage->loadMultiple(), function ($action) use ($entity_type) {
+      return $action->getType() == $entity_type;
+    });
+
+    if (empty($actions)) {
+      return;
+    }
+
+    $form_state->set('entity_reference_' . $field_definition->getName(), $context['items']);
```

Comment:
What about using the setWidgetState() / getWidgetState() functionality? The current implementation could break when the same field is attached multiple times in the form array